### PR TITLE
Move samples & label validation to the distributor.

### PR DESCRIPTION
**Pull request by tomwilkie**
Move samples & label validation to the distributor - most validation is stateless, so not point in doing it 3 times with all the rpc overhead.  Should move a small amount of load of the ingester.

In doing this we also separate out label validation from sample validation, hopefully reducing the amount of times we have to validate a set of strings.

One downside is that the distributors don't know the CreationGracePeriod, so this flag is duplicated.  These other flags have been deprecated and moved, and will need adding to the distributor during upgrade:
* `-ingester.reject-old-samples` -> `-validation.reject-old-samples`
* `-ingester.reject-old-samples.max-age` -> `-validation.reject-old-samples.max-age`
* `-ingester.validation.max-length-label-name` -> `-validation.max-length-label-name`
* `-ingester.validation.max-length-label-value` -> `-validation.max-length-label-value`
* `-ingester.max-label-names-per-series` -> `-validation.max-label-names-per-series`

TODO:
- [x] Improve and make consistent all the errors & metrics. (Byran did it!)
- [x] Consider not rejecting entire batch if one series / metric fails validation. (Done - this is preserving the ingester behaviour)

### CODE REVIEW

File: pkg/ingester/ingester.go [hunk 1]
```diff
@@ -109,6 +96,13 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	f.DurationVar(&cfg.MaxChunkAge, "ingester.max-chunk-age", 12*time.Hour, "Maximum chunk age before flushing.")
 	f.IntVar(&cfg.ConcurrentFlushes, "ingester.concurrent-flushes", DefaultConcurrentFlush, "Number of concurrent goroutines flushing to dynamodb.")
 	f.StringVar(&cfg.ChunkEncoding, "ingester.chunk-encoding", "1", "Encoding version to use for chunks.")
+
+	// DEPRECATED, no-op
+	f.Bool("ingester.reject-old-samples", false, "DEPRECATED. Reject old samples.")
```
Comment (bboreham) [inline pkg/ingester/ingester.go]: should probably say something in the PR description about upgrade procedure - need to enable the checks in the new distributor before upgrading the ingester

File: pkg/ingester/user_state.go [hunk 2]
```diff
@@ -168,7 +169,7 @@ func (u *userState) getSeries(metric model.Metric, cfg *UserStatesConfig) (model
 		return fp, nil, httpgrpc.Errorf(http.StatusTooManyRequests, "per-user series limit (%d) exceeded", cfg.MaxSeriesPerUser)
 	}
 
-	metricName, err := util.ExtractMetricNameFromMetric(metric)
+	metricName, err := extract.MetricNameFromMetric(metric)
```
Comment (bboreham) [inline pkg/ingester/user_state.go]: Why isn't this change part of the previous commit?

File: pkg/ingester/ingester.go (hunk 1 above)
Comment (tomwilkie) [reply pkg/ingester/ingester.go]: Done.

File: pkg/ingester/user_state.go (hunk 2 above)
Comment (tomwilkie) [reply pkg/ingester/user_state.go]: Oh probably got mixed up in various rebases; should be fixed now.

**Comment by tomwilkie** [comment]
Thanks!  Should all be addressed now.

File: pkg/util/validation/validate.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	RejectOldSamples       bool
+	RejectOldSamplesMaxAge time.Duration
+	CreationGracePeriod    time.Duration
+}
+
+// RegisterFlags registers a set of command line flags for setting options regarding sample validation at ingestion time.
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+	f.IntVar(&cfg.MaxLabelNameLength, "validation.max-length-label-name", 1024, "Maximum length accepted for label names")
+	f.IntVar(&cfg.MaxLabelValueLength, "validation.max-length-label-value", 2048, "Maximum length accepted for label value. This setting also applies to the metric name")
+	f.IntVar(&cfg.MaxLabelNamesPerSeries, "validation.max-label-names-per-series", 20, "Maximum number of label names per series.")
```
Comment (bboreham) [inline pkg/util/validation/validate.go]: btw I now think this default is too low - we've got it set to 30

Comment (tomwilkie) [reply pkg/util/validation/validate.go]: Might as well bump it in this PR then, give us a sec.

**Comment by bboreham** [comment]
I think, when upgrading a running system to include this change, you should ensure the distributor is fully upgraded before starting the ingester rollout.

**Comment by bboreham** [comment]
Upgrade instructions should mention `-validation.create-grace-period`

**Pull request merged**